src/H5FDsubfiling/H5subfiling_common.c

```diff
@@ -801,44 +730,39 @@ init_subfiling(const char *base_filename, uint64_t file_id, H5FD_subfiling_param
     char                *file_basename = NULL;
     char                *subfile_dir   = NULL;
     char                *prefix_env    = NULL;
+    char                *env_value     = NULL;
     int                  mpi_rank;
     int                  mpi_size;
     int                  mpi_code;
     herr_t               ret_value = SUCCEED;
 
+    FUNC_ENTER_PACKAGE
+
     assert(context_id_out);
 
     if (MPI_SUCCESS != (mpi_code = MPI_Comm_rank(comm, &mpi_rank)))
-        H5_SUBFILING_MPI_GOTO_ERROR(FAIL, "MPI_Comm_rank failed", mpi_code);
+        HMPI_GOTO_ERROR(FAIL, "MPI_Comm_rank failed", mpi_code);
     if (MPI_SUCCESS != (mpi_code = MPI_Comm_size(comm, &mpi_size)))
-        H5_SUBFILING_MPI_GOTO_ERROR(FAIL, "MPI_Comm_size failed", mpi_code);
+        HMPI_GOTO_ERROR(FAIL, "MPI_Comm_size failed", mpi_code);
 
     /* Use the file's index to create a new subfiling context ID */
-    if ((context_id = H5_new_subfiling_object_id(SF_CONTEXT)) < 0)
-        H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "couldn't create new subfiling context ID");
+    if ((context_id = H5FD__subfiling_new_object_id(SF_CONTEXT)) < 0)
+        HGOTO_ERROR(H5E_VFL, H5E_CANTGET, FAIL, "couldn't create new subfiling context ID");
 
     /* Create a new subfiling context object with the created context ID */
-    if (NULL == (new_context = H5_get_subfiling_object(context_id)))
-        H5_SUBFILING_GOTO_ERROR(H5E_VFL, H5E_CANTINIT, FAIL, "couldn't create new subfiling object");
-    new_context->sf_context_id = -1;
-    new_context->topology      = NULL;
-    new_context->sf_msg_comm   = MPI_COMM_NULL;
-    new_context->sf_data_comm  = MPI_COMM_NULL;
-    new_context->sf_eof_comm   = MPI_COMM_NULL;
-    new_context->sf_node_comm  = MPI_COMM_NULL;
-    new_context->sf_group_comm = MPI_COMM_NULL;
```

Comment:
I'd generally prefer all the initialization and de-initialization code removed in this PR to stay intact. While removing these shouldn't generally pose a problem at the moment due to the current ordering of functions that are called, I can't really predict the future and this could easily turn into an `MPI_Comm_free` on an uninitialized communicator.

Reply:
These initializations are still intact, they were duplicated by later code (around lines 860-870 now), so I removed this copy.

Reply:
The main issue I was trying to get to is that if these fields aren't initialized right after the context object is allocated, there are several errors that could occur in between and the logic could hit the cleanup code with a partially-initialized context object.

Reply:
Agree - now that I understand the intent of this routine better, I'll factor the context initialization back into a subroutine again.  That can bring the initialization back to following the allocation, while avoiding the duplication.  

Probably have it done later tonight.

Reply:
Done